File: apps/web/src/lib/resample.ts
import type { Candle } from "@kronos/shared";

/**
 * Aggregate candles to a higher timeframe (mirror of apps/api data/resample.py).
 * Trading on 4h / 1d instead of 1h is the biggest measured improvement — far
 * fewer bars means far less fee/slippage drag.
 */
export type Timeframe = "1h" | "4h" | "1d";

export const TIMEFRAME_FACTOR: Record<Timeframe, number> = { "1h": 1, "4h": 4, "1d": 24 };
export const TIMEFRAMES: Timeframe[] = ["1h", "4h", "1d"];

export function resampleCandles(candles: Candle[], factor: number): Candle[] {
  if (factor <= 1) return candles;
  const out: Candle[] = [];
  for (let i = 0; i + factor <= candles.length; i += factor) {
    const w = candles.slice(i, i + factor);
    out.push({
      openTime: w[0]!.openTime,
      open: w[0]!.open,
      high: Math.max(...w.map((c) => c.high)),
      low: Math.min(...w.map((c) => c.low)),
      close: w[w.length - 1]!.close,
      volume: w.reduce((a, c) => a + c.volume, 0),
      closed: true,
    });
  }
  return out;
}

File: apps/api/src/data/resample.py
"""Resample candles to a higher timeframe.

Trading on higher timeframes (4h / 1d) instead of hourly is the single biggest,
cheapest improvement we measured: fewer bars => far less turnover => much lower
fee/slippage drag (on real ETH it flipped the full pipeline from -2.55% hourly
to break-even daily). Aggregation uses only closed candles (invariant 1).
"""

from __future__ import annotations

from typing import Sequence

from src.data.ohlcv import Candle, interval_to_ms


def resample(candles: Sequence[Candle], factor: int) -> list[Candle]:
    """Aggregate every `factor` consecutive candles into one (OHLCV)."""
    if factor < 1:
        raise ValueError("factor must be >= 1")
    if factor == 1:
        return list(candles)
    out: list[Candle] = []
    for i in range(0, len(candles) - factor + 1, factor):
        window = candles[i : i + factor]
        out.append(
            Candle(
                open_time=window[0].open_time,
                open=window[0].open,
                high=max(c.high for c in window),
                low=min(c.low for c in window),
                close=window[-1].close,
                volume=sum(c.volume for c in window),
                closed=all(c.closed for c in window),
            )
        )
    return out


def resample_to(candles: Sequence[Candle], src_interval: str, dst_interval: str) -> list[Candle]:
    """Resample from `src_interval` to a larger `dst_interval` (e.g. 1h -> 1d)."""
    src = interval_to_ms(src_interval)
    dst = interval_to_ms(dst_interval)
    if dst < src or dst % src != 0:
        raise ValueError(f"{dst_interval} must be a positive multiple of {src_interval}")
    return resample(candles, dst // src)
